### PR TITLE
Optimize compute RBF Kernel for Matrix to Matrix case

### PR DESCRIPTION
**Pull request by PetrovKP**
Optimization and refactoring RBF Kernel:
1. Common tbb::thread for block compute GEMM - Xi*Yj, Xi^2, Yj^2, and exp with sum these blocks.
2. Local tls struct for GEMM result and store Yj precompute

### CODE REVIEW

**Comment by PetrovKP** [comment]
@Vika-F , @owerbat , @ShvetsKS Please, look this PR

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i [hunk 1]
```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
...
+            algorithmFPType * mklBuff   = tlsLocal->mklBuff;
+            algorithmFPType * sqrDataA2 = tlsLocal->sqrDataA2;
 
-            ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), startRow, numRowsInBlock);
+            ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), startRow2, nRowsInBlock2);
             DAAL_CHECK_BLOCK_STATUS_THR(mtA2);
             const algorithmFPType * dataA2 = const_cast<algorithmFPType *>(mtA2.get());
 
-            for (size_t i = startRow, ii = 0; i < finishRow; ++i, ++ii)
+            for (size_t i = 0; i < nRowsInBlock2; ++i)
```
Comment (ShvetsKS) [inline algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:215]: would be `gemv` with displaces suitable here?

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i [hunk 2]
```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
...
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
-                for (size_t k = 0; k < nVectors2; ++k)
+                algorithmFPType sqrDataA1 = zero;
+                if (a1 != a2)
                 {
-                    dataR[ii * nVectors2 + k] += (sqrDataA1[i] + sqrDataA2[k]);
-                    dataR[ii * nVectors2 + k] *= coeff;
-                    if (dataR[ii * nVectors2 + k] < Math<algorithmFPType, cpu>::vExpThreshold())
+                    for (size_t j = 0; j < nFeatures; ++j)
```
Comment (ShvetsKS) [inline algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:234]: dot-product can be applied here

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i [hunk 3]
```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
...
+            DAAL_INT ldc = blockSize;
+            Blas<algorithmFPType, cpu>::xxgemm(&trans, &notrans, &nRowsInBlock2, &nRowsInBlock1, (DAAL_INT *)&nFeatures, &negTwo, dataA2, &ldb,
+                                               dataA1, &lda, &zero, mklBuff, &ldc);
+            for (size_t i = 0; i < nRowsInBlock1; ++i)
             {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
-                for (size_t k = 0; k < nVectors2; ++k)
+                algorithmFPType sqrDataA1 = zero;
+                if (a1 != a2)
```
Comment (ShvetsKS) [inline algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:232]: just some thoughts:
we could use this statement as template parameter initialization before call `computeInternalVectorVector`:
ex:
```
if (a1 != a2)
  computeInternalVectorVector<true>(...);
else
  computeInternalVectorVector<false>(...);
```

and during call of specific template instantiation of this func we could avoid this condition by compile time optimizations.

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i [hunk 4]
```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
...
+                else
                 {
-                    dataR[i * nVectors1 + k] = coeff * (diagonal[i] + diagonal[k] - two * dataR[i * nVectors1 + k]);
+                    sqrDataA1 = sqrDataA2[i];
                 }
-
-                dataR[i * nVectors1 + i] = zero;
-                daal::internal::Math<algorithmFPType, cpu>::vExp(i + 1, dataR + i * nVectors1, dataR + i * nVectors1);
+                algorithmFPType * dataRBlock   = &dataR[i * nVectors2 + startRow2];
+                algorithmFPType * mklBuffBlock = &mklBuff[i * blockSize];
```
Comment (ShvetsKS) [inline algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:244]: const algorithmFPType * const mklBuffBlock

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i [hunk 5]
```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
...
-                for (size_t k = 0; k < i; ++k)
+                else
                 {
-                    dataR[i * nVectors1 + k] = coeff * (diagonal[i] + diagonal[k] - two * dataR[i * nVectors1 + k]);
+                    sqrDataA1 = sqrDataA2[i];
                 }
-
-                dataR[i * nVectors1 + i] = zero;
-                daal::internal::Math<algorithmFPType, cpu>::vExp(i + 1, dataR + i * nVectors1, dataR + i * nVectors1);
+                algorithmFPType * dataRBlock   = &dataR[i * nVectors2 + startRow2];
```
Comment (ShvetsKS) [inline algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:243]: if  `vExp` allows, please use const qualifier.

**Comment by PetrovKP** [comment]
> what performance gain was obtained?

x2-3 🛩️

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i (hunk 1 above)
Comment (PetrovKP) [reply algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i:231]: This operation does not fit. `A[NxP] (?) A[NxP] = B[Nx1]`

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i (hunk 2 above)
Comment (PetrovKP) [reply algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i]: Good comment, I applied

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i (hunk 3 above)
Comment (PetrovKP) [reply algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i]: I redid this `for` a bit. There is no more of this condition.

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i (hunk 4 above)
Comment (PetrovKP) [reply algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i]: Applied

File: algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i (hunk 5 above)
Comment (PetrovKP) [reply algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i]: Applied

**Pull request merged**